rust/lance-core/src/utils/mask/nullable.rs

```diff
@@ -0,0 +1,457 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use deepsize::DeepSizeOf;
+
+use super::{RowIdMask, RowIdTreeMap};
+
+/// A set of row ids, with optional set of nulls.
+///
+/// This is often a result of a filter, where `selected` represents the rows that
+/// passed the filter, and `nulls` represents the rows where the filter evaluated
+/// to null. For example, in SQL `NULL > 5` evaluates to null. This is distinct
+/// from being deselected to support proper three-valued logic for NOT.
+/// (`NOT FALSE` is TRUE, `NOT TRUE` is FALSE, but `NOT NULL` is NULL.)
+#[derive(Clone, Debug, Default, DeepSizeOf)]
+pub struct NullableRowIdSet {
+    selected: RowIdTreeMap,
+    nulls: RowIdTreeMap,
+}
+
+impl NullableRowIdSet {
+    /// Create a new RowSelection from selected rows and null rows.
+    ///
+    /// `nulls` may have overlap with `selected`. Rows in `nulls` are considered NULL,
+    /// even if they are also in `selected`.
+    pub fn new(selected: RowIdTreeMap, nulls: RowIdTreeMap) -> Self {
+        Self { selected, nulls }
+    }
+
+    pub fn with_nulls(mut self, nulls: RowIdTreeMap) -> Self {
+        self.nulls = nulls;
+        self
+    }
+
+    /// Create an empty selection. Alias for [Default::default]
+    pub fn empty() -> Self {
+        Default::default()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.selected.is_empty()
+    }
+
+    /// Check if a row_id is selected (TRUE)
+    pub fn selected(&self, row_id: u64) -> bool {
+        self.selected.contains(row_id) && !self.nulls.contains(row_id)
+    }
+
+    /// Get the selected rows (TRUE or NULL)
+    pub fn selected_rows(&self) -> &RowIdTreeMap {
+        &self.selected
+    }
```

Comment:
It's a little unexpected that `selected` checks for `TRUE` but `selected_rows` is both `TRUE` and `NULL`.  Maybe change the name of one of these two?

Reply:
I'm just going to remove this, as it's not that useful. The places where we use it are just in tests and for those we really want `.true_rows()`.